### PR TITLE
Ability to save answers as a PDF

### PR DESCRIPTION
**Pull request by MebinAbraham**
### What is the context of this PR?

- Adds the ability to save the answers as a PDF on the view response page.
- Since the print style sheet need to live in runner, I have added a script to pull the latest style sheet from the CDN when loading the design system, otherwise, the styles may become out of date and the rendered PDF may not look as expected.

### How to review
Use `test_view_submitted_response.json` schema to ensure you are able to download the answers as a PDF. 

- Deploy this to GCP or run it via Docker to ensure it works as expected under Debian.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

File: tests/app/views/handlers/test_view_submitted_response_pdf.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        self.schema = QuestionnaireSchema({"post_submission": {"view_response": True}})
+        self.language = "en"
+
+    def set_submitted_at(self, submitted_at):
+        self.storage = Mock()
+        self.storage.get_user_data = Mock(return_value=("{}", 1, submitted_at))
+
+    def test_pdf_not_downloadable(self):
+        submitted_at = datetime.now(timezone.utc) - timedelta(minutes=60)
```
Comment (jamesnunn) [inline tests/app/views/handlers/test_view_submitted_response_pdf.py:25]: Only splitting hairs but would you not test a boundary condition like -46 mins?

File: tests/app/views/handlers/test_view_submitted_response_pdf.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        self.questionnaire_store = QuestionnaireStore(self.storage)
+
+        self.questionnaire_store.set_metadata(
+            {"schema_name": "test_view_submitted_response"}
+        )
+        pdf = ViewSubmittedResponsePDF(
+            self.schema, self.questionnaire_store, self.language
+        )
+
+        self.assertEqual(pdf.filename, f"test_view_submitted_response.pdf")
```
Comment (jamesnunn) [inline tests/app/views/handlers/test_view_submitted_response_pdf.py:48]: f string has no replacement targets here so I guess it's not needed?

File: tests/app/views/handlers/test_view_submitted_response_pdf.py (hunk 1 above)
Comment (MebinAbraham) [reply tests/app/views/handlers/test_view_submitted_response_pdf.py]: Fair enough, updated.

File: app/questionnaire/questionnaire_schema.py [hunk 3]
```diff
@@ -209,6 +209,12 @@ def get_post_submission(self) -> ImmutableDict:
         schema: ImmutableDict = self.json.get("post_submission", ImmutableDict({}))
         return schema
 
+    @property
+    def is_view_submitted_response_enabled(self) -> bool:
```
Comment (parimooa) [inline app/questionnaire/questionnaire_schema.py:213]: why is this  `property` instead of method?

File: app/views/handlers/view_submitted_response_pdf.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        return f"{self._metadata['schema_name']}.pdf"
+
+    def get_pdf(self) -> io.BytesIO:
+        content_as_bytes = pdfkit.from_string(
+            input=self.get_rendered_html(),
+            output_path=None,
+            css=f'{current_app.config["PRINT_STYLE_SHEET_FILE_PATH"]}/print.css',
+            options={
+                "quiet": "",
+                "dpi": 365,
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/views/handlers/view_submitted_response_pdf.py:40]: Do you mean the `dpi` or the literal options?.

File: tests/app/views/handlers/test_view_submitted_response_pdf.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+from datetime import datetime, timedelta, timezone
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply tests/app/views/handlers/test_view_submitted_response_pdf.py:1]: To keep it consistent with the existing `test_view_submitted_response.py`, which, this has overlap with and was written using unit tests. Would be happy to the new one and existing to pytest.

File: app/questionnaire/questionnaire_schema.py (hunk 3 above)
Comment (MebinAbraham) [reply app/questionnaire/questionnaire_schema.py:213]: Why does it need to be a method?. I see it as a property since we do not pass any arguments, which is consistent with our other `is_` functions on questionnaire schema class. It uses method when you have to pass args, and property when you don't.

File: tests/app/views/handlers/test_view_submitted_response_pdf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply tests/app/views/handlers/test_view_submitted_response_pdf.py:1]: Actually my bad, I was looking at the wrong file, I was looking at the integration test not this, however even for this the existing one is unittest but given it is only 1 existing test, ill migrate them to pytest so it leaves us in a better shape.

File: app/questionnaire/questionnaire_schema.py (hunk 3 above)
Comment (parimooa) [reply app/questionnaire/questionnaire_schema.py:213]: Fair enough ? could it be a `@cached_property` like other properties to be consistent ? i am not sure.

Comment (MebinAbraham) [reply app/questionnaire/questionnaire_schema.py]: Good point, changed to `@cached_property`.

File: app/views/handlers/view_submitted_response_pdf.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/views/handlers/view_submitted_response_pdf.py]: Made options a class attribute.

File: tests/app/views/handlers/test_view_submitted_response_pdf.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+from unittest.mock import Mock
```
Comment (parimooa) [inline tests/app/views/handlers/test_view_submitted_response_pdf.py:2]: Dont see `Mock` getting used in here ?

File: tests/app/views/handlers/test_view_submitted_response_pdf.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+import pytest
+
+from app.data_models import QuestionnaireStore
+from app.questionnaire.questionnaire_schema import QuestionnaireSchema
+from app.views.handlers.view_submitted_response import ViewSubmittedResponseExpired
+from app.views.handlers.view_submitted_response_pdf import ViewSubmittedResponsePDF
+from tests.app.views.handlers.conftest import set_storage_data
```
Comment (parimooa) [inline tests/app/views/handlers/test_view_submitted_response_pdf.py:10]: this could be `from .conftest import set_storage_data` ?

File: tests/app/views/handlers/conftest.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+from unittest.mock import Mock
```
Comment (parimooa) [inline tests/app/views/handlers/conftest.py:1]: Consider using [monkeypatch](https://docs.pytest.org/en/6.2.x/monkeypatch.html), i will leave it up to you.

File: tests/app/views/handlers/test_view_submitted_response_pdf.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+from datetime import datetime, timedelta, timezone
+from unittest.mock import Mock
+
+import pytest
+
+from app.data_models import QuestionnaireStore
+from app.questionnaire.questionnaire_schema import QuestionnaireSchema
```
Comment (parimooa) [inline tests/app/views/handlers/test_view_submitted_response_pdf.py:7]: Also `QuestionnaireSchema` is not being used anywhere

File: tests/app/views/handlers/test_view_submitted_response_pdf.py (hunk 6 above)
Comment (MebinAbraham) [reply tests/app/views/handlers/test_view_submitted_response_pdf.py]: Removed, we really need to get those pylint cards in, as lint should fail for these but don't atm 😞 .

File: tests/app/views/handlers/conftest.py (hunk 8 above)
Comment (MebinAbraham) [reply tests/app/views/handlers/conftest.py:1]: Haven't actually used `monkeypatch` for test yet, but we could consider that, atm I used `Mock` as that is what we do across runner.

File: tests/app/views/handlers/test_view_submitted_response_pdf.py (hunk 7 above)
Comment (parimooa) [reply tests/app/views/handlers/test_view_submitted_response_pdf.py]: coming from AWS lambda , it was a pain to have relative imports. very weird. If we use relative across let it be like that. TBH iam not fussed about it.

File: tests/integration/routes/test_view_submitted_response_pdf.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        self.post()
+
+        # When I try to download my answer
+        self.get("/submitted/download-pdf")
+
+        # Then a 404 page is returned
+        self.assertStatusNotFound()
+
+    def test_download_when_submitted_response_enabled_but_not_expired(self):
+        # Given I launch and complete a questionnaire that does not have view-submitted-response enabled
```
Comment (parimooa) [inline tests/integration/routes/test_view_submitted_response_pdf.py:23]: this comment contradicts the test , is `view-submitted-response` enabled or disabled ?

File: tests/integration/routes/test_view_submitted_response_pdf.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        self._launch_and_complete_questionnaire()
+
+        # When I try to download my answer
+        self.get(self.VIEW_RESPONSE_PAGE_URL)
+        self.get(self.get_download_button().attrs["href"])
+
+        # Then a the download is successful
```
Comment (parimooa) [inline tests/integration/routes/test_view_submitted_response_pdf.py:30]: typo ?

File: app/routes/questionnaire.py [hunk 12]
```diff
@@ -393,14 +395,34 @@ def get_view_submitted_response(schema, questionnaire_store):
...
-        template="view-submitted-response",
-        content=view_submitted_response.get_context(),
-        page_title=view_submitted_response.get_page_title(),
+    return view_submitted_response.get_rendered_html()
+
+
+@post_submission_blueprint.route("download-pdf", methods=["GET"])
+@with_questionnaire_store
+@with_schema
+def get_view_submitted_response_pdf(
```
Comment (parimooa) [inline app/routes/questionnaire.py:407]: consider doc strings

File: app/routes/questionnaire.py [hunk 13]
```diff
@@ -393,14 +395,34 @@ def get_view_submitted_response(schema, questionnaire_store):
...
+    schema: QuestionnaireSchema, questionnaire_store: QuestionnaireStore
+):
```
Comment (parimooa) [inline app/routes/questionnaire.py:409]: return type missing

File: app/views/handlers/view_submitted_response_pdf.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+from app.data_models import QuestionnaireStore
+from app.questionnaire import QuestionnaireSchema
+from app.views.handlers.view_submitted_response import (
+    ViewSubmittedResponse,
+    ViewSubmittedResponseExpired,
+)
+
+
+class ViewSubmittedResponsePDF(ViewSubmittedResponse):
+    mimetype = "application/pdf"
```
Comment (parimooa) [inline app/views/handlers/view_submitted_response_pdf.py:15]: consider doc strings

File: tests/integration/routes/test_view_submitted_response_pdf.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+        # Check filename is set as expected
+        assert (
+            "filename=test_view_submitted_response.pdf"
+            in self.last_response_headers["Content-Disposition"]
+        )
+
+        # Check content length is reasonable.
+        # This is given some leeway as it can change with DS changes.
+        self.assertGreater(self.last_response.content_length, 16500)
```
Comment (parimooa) [inline tests/integration/routes/test_view_submitted_response_pdf.py:47]: why not `assert self.last_response.content_length > 16500 ,"PDF content is less than 16500"`

File: tests/integration/routes/test_view_submitted_response_pdf.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+        # When I try to download my answer
+        self.get(self.VIEW_RESPONSE_PAGE_URL)
+        self.get(self.get_download_button().attrs["href"])
+
+        # Then a the download is successful
+        self.assertStatusOK()
+
+        # Check response content type is PDF
+        self.assertEqual(self.last_response.content_type, "application/pdf")
```
Comment (parimooa) [inline tests/integration/routes/test_view_submitted_response_pdf.py:34]:
```suggestion
       assert self.last_response.content_type == "application/pdf" ," Response type is not of 'PDF' "
```

File: app/routes/questionnaire.py (hunk 12 above)
Comment (MebinAbraham) [reply app/routes/questionnaire.py:407]: Have added doc string. We should have a wider convo with the type on the style guide for doc strings and when we should have them etc. The doc strings we have mostly conform to `Sphinx` format, but I personally prefer the Google's style guide and it not as visually dense. We should discuss the options in a tech session and put together or follow a style guide.

File: tests/integration/routes/test_view_submitted_response_pdf.py (hunk 16 above)
Comment (MebinAbraham) [reply tests/integration/routes/test_view_submitted_response_pdf.py:34]: https://github.com/ONSdigital/eq-questionnaire-runner/pull/685#discussion_r731835329
